File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<pair<int,int>,int>>q;
        int m = grid.size();
        int n = grid[0].size();
        int fresh = 0;
        for(int i = 0 ; i < m ; i++){
            for(int j = 0 ; j < n ; j++){
                if(grid[i][j] == 2){
                    q.push({{i,j},0});
                }
                if(grid[i][j] == 1){
                    fresh++;
                }
            }
        }
        int latest = 0;
        while(!q.empty()){
            auto rotts = q.size();
            while(rotts--){
                auto recentRott = q.front();
                pair<int,int>rottPos = recentRott.first;
                latest = max(latest,recentRott.second);
                int i = rottPos.first;
                int j = rottPos.second;
                if(i < m && j+1 < n) {
                    if(grid[i][j+1] == 1){
                        fresh--;
                        q.push({{i,j+1},recentRott.second+1});
                        grid[i][j+1] = 2;
                    }
                }
                if(i+1 < m && j < n) {
                    if(grid[i+1][j] == 1){
                        fresh--;
                        q.push({{i+1,j},recentRott.second+1});
                        grid[i+1][j] = 2;
                    }
                }
                if(i < m && j-1 >= 0) {
                    if(grid[i][j-1] == 1){
                        fresh--;
                        q.push({{i,j-1},recentRott.second+1});
                        grid[i][j-1] = 2;
                    }
                }
                if(i-1 >= 0 && j < n) {
                    if(grid[i-1][j] == 1){
                        fresh--;
                        q.push({{i-1,j},recentRott.second+1});
                        grid[i-1][j] = 2;
                    }
                }
                q.pop();
            }
        }
        if(fresh > 0) return -1;
        return latest;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int r = grid.length;
        int c = grid[0].length;
        Queue<int[]> q = new LinkedList<>();
        int good = 0, time = 0;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {   // fixed cols → c
                if (grid[i][j] == 2) q.add(new int[]{i, j});
                else if (grid[i][j] == 1) good++;
            }
        }
        int[][] d = {{1,0}, {-1,0}, {0,1}, {0,-1}};
        while (!q.isEmpty() && good > 0) {
            int s = q.size();
            for (int i = 0; i < s; i++) {
                int[] cur = q.poll();
                for (int[] dir : d) {
                    int nr = cur[0] + dir[0];
                    int nc = cur[1] + dir[1];
                    if (nr >= 0 && nr < r && nc >= 0 && nc < c && grid[nr][nc] == 1) {
                        grid[nr][nc] = 2;
                        q.add(new int[]{nr, nc});
                        good--;
                    }
                }
            }
            time++;
        }
        return good == 0 ? time : -1;
    }
}